web/virtualhostresource.go

```diff
@@ -100,50 +100,100 @@ func restAddVirtualHost(w *rest.ResponseWriter, r *rest.Request, client *node.Co
 
 // restRemoveVirtualHost removes a vhost name from provided service and endpoint. Parameters are defined in path.
 func restRemoveVirtualHost(w *rest.ResponseWriter, r *rest.Request, client *node.ControlClient) {
-	serviceID, err := url.QueryUnescape(r.PathParam("serviceId"))
+
+	service, application, vhostname, err := getVHostContext(r, client)
 	if err != nil {
-		glog.Errorf("Failed getting serviceId: %v", err)
-		restBadRequest(w, err)
+		restServerError(w, err)
 		return
 	}
-	application, err := url.QueryUnescape(r.PathParam("application"))
+
+	err = service.RemoveVirtualHost(application, vhostname)
 	if err != nil {
-		glog.Errorf("Failed getting application: %v", err)
-		restBadRequest(w, err)
+		glog.Errorf("Unexpected error removing vhost, %s, from service (%s): %v", vhostname, service.ID, err)
+		restServerError(w, err)
 		return
 	}
 
-	hostname, err := url.QueryUnescape(r.PathParam("name"))
+	var unused int
+	err = client.UpdateService(*service, &unused)
 	if err != nil {
-		glog.Errorf("Failed getting hostname: %v", err)
-		restBadRequest(w, err)
+		glog.Errorf("Unexpected error removing vhost, %s, from service (%s): %v", vhostname, service.ID, err)
+		restServerError(w, err)
 		return
 	}
 
+	restSuccess(w)
+}
+
+
+// return serviceID, application and vhostname from the URL path
+func getVHostContext(r *rest.Request,  client *node.ControlClient) (*service.Service, string, string, error){
+	serviceID, err := url.QueryUnescape(r.PathParam("serviceId"))
+	if err != nil {
+		glog.Errorf("Failed getting serviceID: %v", err)
+		return nil, "", "", err
+	}
+
 	var service service.Service
 	err = client.GetService(serviceID, &service)
 	if err != nil {
 		glog.Errorf("Unexpected error getting service (%s): %v", serviceID, err)
-		restServerError(w, err)
+		return nil, "", "", err
+	}
+
+	application, err := url.QueryUnescape(r.PathParam("application"))
+	if err != nil {
+		glog.Errorf("Failed getting application: %v", err)
+		return nil, "", "", err
+	}
+
+	vhostname, err := url.QueryUnescape(r.PathParam("name"))
+	if err != nil {
+		glog.Errorf("Failed getting hostname: %v", err)
+		return nil, "", "", err
+	}
+	return &service, application, vhostname, nil
+
+}
+
+// json object for enabling/disabling a virtual host
+type virtualHostEnable struct {
+	Enable bool
+}
+
+// restVirtualHostEnable enables or disables a virtual host endpoint
+func restVirtualHostEnable(w *rest.ResponseWriter, r *rest.Request, client *node.ControlClient) {
+	var request virtualHostEnable
+	err := r.DecodeJsonPayload(&request)
+	if err != nil {
+		restBadRequest(w, err)
 		return
 	}
+	glog.V(1).Infof("Enable VHOST with %s %#v", r.URL.Path, request)
 
-	err = service.RemoveVirtualHost(application, hostname)
+	service, application, vhostname, err := getVHostContext(r, client)
 	if err != nil {
-		glog.Errorf("Unexpected error removing vhost, %s, from service (%s): %v", hostname, serviceID, err)
+		restServerError(w, err)
```

Comment:
Our REST error codes in general are not that great, but this is another example where we're conflating unexpected server-side errors with user-input errors. If the URL is valid, but either the serviceid, application/endpoint name or vhost name is undefined, then that should really be a 404 (though we don't really return 404s today).  If the server encountered some other unexpected error, then a 500 (server-error) is the correct response.

Since the current REST backend treats all not-found conditions as 500, then we should probably leave this as is for now, but see note in EnableVirtualHost()